### ISSUE TITLE
Hardcoded path in glibc causes issues when /tmp is mounted noexec

### ISSUE THREAD

**Issue by EduardBaer** [Homebrew/homebrew-core#111982]
### `brew gist-logs <formula>` link OR `brew config` AND `brew doctor` output

```shell
brew config

HOMEBREW_VERSION: 3.6.3-28-g7b00b2b
ORIGIN: https://github.com/Homebrew/brew.git
HEAD: 7b00b2b1d70e045a5ad2fd952ed9b4b0dc2abd49
Last commit: 10 hours ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: dffc2e43e1ff188970ebf0dce87a2f26a279ce75
Core tap last commit: 38 minutes ago
Core tap branch: master
HOMEBREW_PREFIX: /usr/home/lvlnzvirta/.linuxbrew
HOMEBREW_REPOSITORY: /usr/home/lvlnzvirta/.linuxbrew/Homebrew
HOMEBREW_CELLAR: /usr/home/lvlnzvirta/.linuxbrew/Cellar
HOMEBREW_CASK_OPTS: []
HOMEBREW_MAKE_JOBS: 2
HOMEBREW_TEMP: /usr/home/lvlnzvirta/.software-tmp
Homebrew Ruby: 2.6.8 => /usr/home/lvlnzvirta/.linuxbrew/Homebrew/Library/Homebrew/vendor/portable-ruby/2.6.8_1/bin/ruby
CPU: dual-core 64-bit sandybridge
Clang: N/A
Git: 2.30.2 => /bin/git
Curl: 7.74.0 => /bin/curl
Kernel: Linux 5.10.0-18-amd64 x86_64 GNU/Linux
OS: Debian GNU/Linux 11 (bullseye) (bullseye)
Host glibc: 2.31
/usr/bin/gcc: 10.2.1
/usr/bin/ruby: 2.7.4
glibc: N/A
gcc@11: N/A
gcc: N/A
xorg: N/A

brew doctor

Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: Your Homebrew's prefix is not /home/linuxbrew/.linuxbrew.
Some of Homebrew's bottles (binary packages) can only be used with the default
prefix (/home/linuxbrew/.linuxbrew).
You will encounter build failures with some formulae.
Please create pull requests instead of asking for help on Homebrew's GitHub,
Twitter or any other official channels. You are responsible for resolving
any issues you experience while you are running this
unsupported configuration.
```

### Verification

- [X] I ran `brew update` and am still able to reproduce my issue.
- [X] I have resolved all warnings from `brew doctor` and that did not fix my problem.

### What were you trying to do (and why)?

I tried to install packages with a `glibc` dependency. The issue also occurs when trying to install `glibc` only.

### What happened (include all command output)?

`glibc` installation fails when `/tmp` is mounted with `noexec`. Setting `HOMEBREW_TEMP` doesn't solve the issue.

```
$ brew install glibc
- HOMEBREW_CELLAR: /home/linuxbrew/.linuxbrew/Cellar (yours is /usr/home/lvlnzvirta/.linuxbrew/Cellar)
- HOMEBREW_PREFIX: /home/linuxbrew/.linuxbrew (yours is /usr/home/lvlnzvirta/.linuxbrew)
==> Downloading https://github.com/Homebrew/glibc-bootstrap/releases/download/1.0.0/bootstrap-binutils-2.38.tar.gz
Already downloaded: /usr/home/lvlnzvirta/.cache/Homebrew/downloads/9daccbc8162c5702590e1405c1fb9c4824aee8c29b607f4ae0f77e06cb827036--bootstrap-binutils-2.38.tar.gz
==> Downloading https://github.com/Homebrew/glibc-bootstrap/releases/download/1.0.0/bootstrap-bison-3.8.2.tar.gz
Already downloaded: /usr/home/lvlnzvirta/.cache/Homebrew/downloads/e8b2b31c87e86aae9b340b3f236f622d7bfa64d68a968b9a27879cd44f3868b2--bootstrap-bison-3.8.2.tar.gz
==> Downloading https://github.com/Homebrew/glibc-bootstrap/releases/download/1.0.0/bootstrap-gawk-5.1.1.tar.gz
Already downloaded: /usr/home/lvlnzvirta/.cache/Homebrew/downloads/69ea20c522780dd7bb9eca0b720d1b8d5b0e725910e014e46d8c42278da2e25c--bootstrap-gawk-5.1.1.tar.gz
==> Downloading https://github.com/Homebrew/glibc-bootstrap/releases/download/1.0.0/bootstrap-gcc-9.5.0.tar.gz
Already downloaded: /usr/home/lvlnzvirta/.cache/Homebrew/downloads/d798bf0a43adeacfa8eabcd3be7716c504985349d6ec35b2aa6cc131df924e85--bootstrap-gcc-9.5.0.tar.gz
==> Downloading https://github.com/Homebrew/glibc-bootstrap/releases/download/1.0.0/bootstrap-make-4.3.tar.gz
Already downloaded: /usr/home/lvlnzvirta/.cache/Homebrew/downloads/d2ccb4e6c70f07e3d6240650c09c5ebc7e93a9c0e6db5b124ce9d518e30358e5--bootstrap-make-4.3.tar.gz
==> Downloading https://github.com/Homebrew/glibc-bootstrap/releases/download/1.0.0/bootstrap-python3-3.9.13.tar.gz
Already downloaded: /usr/home/lvlnzvirta/.cache/Homebrew/downloads/25df16c6cd058f6218b9f9235ce8b0fecf7d350162ffd46aab2de7967d051154--bootstrap-python3-3.9.13.tar.gz
==> Downloading https://github.com/Homebrew/glibc-bootstrap/releases/download/1.0.0/bootstrap-sed-4.8.tar.gz
Already downloaded: /usr/home/lvlnzvirta/.cache/Homebrew/downloads/eeceff913839cbc91176f6c1db39d31391731c20d0bbe8777797c6ca71c51be0--bootstrap-sed-4.8.tar.gz
==> Downloading https://ftp.gnu.org/gnu/glibc/glibc-2.35.tar.gz
Already downloaded: /usr/home/lvlnzvirta/.cache/Homebrew/downloads/329d4c5b7b38da0a5dcb652257748b77e9100bbc4771166317ac852d1376cf82--glibc-2.35.tar.gz
==> ../configure --disable-crypt --disable-debug --disable-dependency-tracking --disable-silent-rules --prefix=/usr/home/lvlnzvirta/.linuxbrew/Cellar/glibc/2.35_1 --sysconfdir=/usr/home/lvlnzvirta/.linuxbrew/etc --enable-obsolete-rpc --
Last 15 lines from /usr/home/lvlnzvirta/.cache/Homebrew/Logs/glibc/01.configure:
--disable-silent-rules
--prefix=/usr/home/lvlnzvirta/.linuxbrew/Cellar/glibc/2.35_1
--sysconfdir=/usr/home/lvlnzvirta/.linuxbrew/etc
--enable-obsolete-rpc
--without-gd
--without-selinux
--with-binutils=/tmp/homebrew/bin
--with-headers=/usr/home/lvlnzvirta/.linuxbrew/opt/linux-headers@5.15/include

checking build system type... x86_64-pc-linux-gnu
checking host system type... x86_64-pc-linux-gnu
checking for gcc... /tmp/homebrew/bin/gcc
checking for suffix of object files... configure: error: in `/usr/home/lvlnzvirta/.software-tmp/glibc-20220929-40083-nmldii/glibc-2.35/build':
configure: error: cannot compute suffix of object files: cannot compile
See `config.log' for more details

READ THIS: https://docs.brew.sh/Troubleshooting

These open issues may also help:
glibc@2.13: disable. https://github.com/Homebrew/homebrew-core/pull/109792
llvm: Fix test when brewed glibc is installed https://github.com/Homebrew/homebrew-core/pull/111921
```

from config.log
```
## ----------- ##
## Core tests. ##
## ----------- ##

configure:2266: checking build system type
configure:2280: result: x86_64-pc-linux-gnu
configure:2300: checking host system type
configure:2313: result: x86_64-pc-linux-gnu
configure:2382: checking for gcc
configure:2409: result: /tmp/homebrew/bin/gcc
configure:2638: checking for C compiler version
configure:2647: /tmp/homebrew/bin/gcc --version >&5
../configure: line 2649: /tmp/homebrew/bin/gcc: Permission denied
configure:2658: $? = 126
configure:2647: /tmp/homebrew/bin/gcc -v >&5
../configure: line 2649: /tmp/homebrew/bin/gcc: Permission denied
configure:2658: $? = 126
configure:2647: /tmp/homebrew/bin/gcc -V >&5
../configure: line 2649: /tmp/homebrew/bin/gcc: Permission denied
configure:2658: $? = 126
configure:2647: /tmp/homebrew/bin/gcc -qversion >&5
../configure: line 2649: /tmp/homebrew/bin/gcc: Permission denied
configure:2658: $? = 126
configure:2663: checking for suffix of object files
configure:2685: /tmp/homebrew/bin/gcc -c   conftest.c >&5
../configure: line 2687: /tmp/homebrew/bin/gcc: Permission denied
configure:2689: $? = 126
configure: failed program was:
| /* confdefs.h */
| #define PACKAGE_NAME "GNU C Library"
| #define PACKAGE_TARNAME "glibc"
| #define PACKAGE_VERSION "(see version.h)"
| #define PACKAGE_STRING "GNU C Library (see version.h)"
| #define PACKAGE_BUGREPORT "https://sourceware.org/bugzilla/"
| #define PACKAGE_URL "https://www.gnu.org/software/glibc/"
| #define PKGVERSION "(GNU libc) "
| #define REPORT_BUGS_TO "<https://www.gnu.org/software/libc/bugs.html>"
| /* end confdefs.h.  */
| 
| int
| main ()
| {
| 
|   ;
|   return 0;
| }
configure:2703: error: in `/usr/home/lvlnzvirta/.software-tmp/glibc-20220929-40083-nmldii/glibc-2.35/build':
configure:2705: error: cannot compute suffix of object files: cannot compile
See `config.log' for more details
```

### What did you expect to happen?

`glibc` should install without errors even when `/tmp` is mounted `noexec`. `HOMEBREW_TEMP` or possibly a new variable should be able to overwrite the currently hardcoded path `/tmp/homebrew`.

I believe this issue has been introduced in https://github.com/Homebrew/homebrew-core/pull/106837

### Step-by-step reproduction instructions (by running `brew` commands)

```shell
- Make sure `/tmp` is mounted with `noexec`
- Set `HOMEBREW_TEMP` to a folder other than `/tmp`
- Run `brew install glibc`
- Notice that `HOMEBREW_TEMP` is ignored and programs are placed in `/tmp/homebrew` for execution
```

**Comment by EduardBaer** [comment]
> This won't work, unfortunately, because it's reliant on these binary packages.
I'm not sure that I understand. Surely those binaries could be placed anywhere as long as the current user has access to that path?

> The alternative would be to build all these components from source in this situation.
Unfortunately `brew install --build-from-source glibc` failed with the same error.

**Comment by carlocab** [comment]
We actually already have [a diagnostic](https://github.com/Homebrew/brew/blob/7b00b2b1d70e045a5ad2fd952ed9b4b0dc2abd49/Library/Homebrew/extend/os/linux/diagnostic.rb#L44-L60) for this, but `glibc` ignores `HOMEBREW_TEMP` because the bootstrap binaries need to be installed into `/tmp/homebrew` in order to function.

If we want to continue this bootstrap setup we would need to duplicate the diagnostic in the `glibc` formula and use it to check whether we can use the bootstrap binaries.

I suspect building `glibc` from source without using the bootstrap binaries doesn't work though, even if the system provides the necessary versions of `binutils`, `gcc`, `python3`, etc (this might be the case for @EduardBaer's system, since `/usr/bin/gcc: 10.2.1`), in part due to this:

https://github.com/Homebrew/homebrew-core/blob/0c9ffc1aae8a2a31f1fe3397a0f0821e0faf6375/Formula/glibc.rb#L148

The other part is that installing `glibc` without using the bootstrap binaries is likely completely untested.

CC @Homebrew/linux, especially @danielnachun

**Comment by danielnachun** [comment]
This should be easy to fix by moving the `binutils` argument to be inside the block that checks if we are on Intel, and making sure that block is conditional on `/tmp` allowing execution. 

I’ll open a PR for this shortly as these fixes are simple. This scenario where `/tmp` cannot be used to execute binaries is not something we will be able to officially support and the user will have to supply sufficiently new binaries.  I’d recommend using `miniconda` to do so in your case.

It should be noted here that this will break a lot of other bootstrapping systems beyond ours that assume things can be executed in `/tmp`. I don’t why your system is configured in this manner but I’d be curious to know why, given how much this can break binary bootstrapping in general.

**Comment by MikeMcQuaid** [comment]
> I’ll open a PR for this shortly as these fixes are simple. This scenario where `/tmp` cannot be used to execute binaries is not something we will be able to officially support and the user will have to supply sufficiently new binaries. I’d recommend using `miniconda` to do so in your case.

I'd like to avoid layering hacks on hacks here. My suggestions (in descending order of preference):
1) we simply do not support this configuration if you want `noexec` `/tmp` and also to have an older system `glibc` and do not modify the formula beyond producing a better error message
2) we drop the whole `glibc-bootstrap` approach and build from source for everyone every time
3) we fail unless you have new enough system versions of these dependencies inside the `glibc` formula

**Comment by danielnachun** [comment]
Definitely have a strong preference for number 1, but I should note that there aren’t really any hacks here, just a simple check for whether `/tmp` allows execution.  I’d still like to make the PR to see how this looks, because I don’t think the changes are unreasonably complex.

**Comment by MikeMcQuaid** [comment]
> but I should note that there aren’t really any hacks here

To be clear, I consider the whole bootstrap-glibc approach (that I suggested!) to be a hack 😂

**Comment by CenterOfEfficiency** [comment]
Any timeline when this isseu will be fixed?
We can not install composer nor node due to this error and our hoster suggest to wait for a bugfix... So highly appreciated

**Comment by CenterOfEfficiency** [comment]
Dear Customer,

we have received your support request and will reply as soon as
possible.
Your Ticket number is: #1634.

Support

 		Center of Efficiency

 		   +49 (0) 7721 68 088 80 

 		   ***@***.*** 

 		   www.centerofefficiency.com
[https://www.centerofefficiency.com/] 

 		   Steppachhalde 43 | 78647 Trossingen | Deutschland 

 Unternehmergesellschaft (haftungsbeschränkt) mit Sitz in Trossingen
|| Amtsgericht Stuttgart || HRB 746635
Geschäftsführer: Dipl.-Wi.-Ing. Bernhard Volz
USt-IdNr.: DE291682477

**Comment by MikeMcQuaid** [comment]
@CenterOfEfficiency There are two fixes you can use today:
- stop mounting `/tmp` `noexec`
- install Homebrew on Linux into the supported path inside `/home/linuxbrew`

**Comment by CenterOfEfficiency** [comment]
Dear Customer,

we have received your support request and will reply as soon as
possible.
Your Ticket number is: #1635.

Support

 		Center of Efficiency

 		   +49 (0) 7721 68 088 80 

 		   ***@***.*** 

 		   www.centerofefficiency.com
[https://www.centerofefficiency.com/] 

 		   Steppachhalde 43 | 78647 Trossingen | Deutschland 

 Unternehmergesellschaft (haftungsbeschränkt) mit Sitz in Trossingen
|| Amtsgericht Stuttgart || HRB 746635
Geschäftsführer: Dipl.-Wi.-Ing. Bernhard Volz
USt-IdNr.: DE291682477

**Comment by CenterOfEfficiency** [comment]
Ok, thanks a lot.
@SMillerDev : no offense meant at all, we really appreciate all your work and efforts.
It's just because the hoster simply says "wait for a bugfix".
We will discuss the solutions mentioned by MikeMcQuaid with the hoster to hopefully get a solution.

**Comment by CenterOfEfficiency** [comment]
Dear Customer,

we have received your support request and will reply as soon as
possible.
Your Ticket number is: #1898.

Support

 		Center of Efficiency

 		   +49 (0) 7721 68 088 80 

 		   ***@***.*** 

 		   www.centerofefficiency.com
[https://www.centerofefficiency.com/] 

 		   Steppachhalde 43 | 78647 Trossingen | Deutschland 

 Unternehmergesellschaft (haftungsbeschränkt) mit Sitz in Trossingen
|| Amtsgericht Stuttgart || HRB 746635
Geschäftsführer: Dipl.-Wi.-Ing. Bernhard Volz
USt-IdNr.: DE291682477

**Comment by CenterOfEfficiency** [comment]
Dear Customer,

we have received your support request and will reply as soon as
possible.
Your Ticket number is: #2048.

Support

 		Center of Efficiency

 		   +49 (0) 7721 68 088 80 

 		   ***@***.*** 

 		   www.centerofefficiency.com
[https://www.centerofefficiency.com/] 

 		   Steppachhalde 43 | 78647 Trossingen | Deutschland 

 Unternehmergesellschaft (haftungsbeschränkt) mit Sitz in Trossingen
|| Amtsgericht Stuttgart || HRB 746635
Geschäftsführer: Dipl.-Wi.-Ing. Bernhard Volz
USt-IdNr.: DE291682477

**Comment by CenterOfEfficiency** [comment]
Dear Customer,

we have received your support request and will reply as soon as
possible.
Your Ticket number is: #2225.

Important: We are closed from 22.12.2022 to 01.01.2023.
Requests will be answered after our return.

Support

 		Center of Efficiency

 		   +49 (0) 7721 68 088 80 

 		   ***@***.*** 

 		   www.centerofefficiency.com
[https://www.centerofefficiency.com/] 

 		   Steppachhalde 43 | 78647 Trossingen | Deutschland 

 Unternehmergesellschaft (haftungsbeschränkt) mit Sitz in Trossingen
|| Amtsgericht Stuttgart || HRB 746635
Geschäftsführer: Dipl.-Wi.-Ing. Bernhard Volz
USt-IdNr.: DE291682477

**Comment by CenterOfEfficiency** [comment]
Dear Customer,

we have received your support request and will reply as soon as
possible.
Your Ticket number is: #2233.

Important: We are closed from 22.12.2022 to 01.01.2023.
Requests will be answered after our return.

Support

 		Center of Efficiency

 		   +49 (0) 7721 68 088 80 

 		   ***@***.*** 

 		   www.centerofefficiency.com
[https://www.centerofefficiency.com/] 

 		   Steppachhalde 43 | 78647 Trossingen | Deutschland 

 Unternehmergesellschaft (haftungsbeschränkt) mit Sitz in Trossingen
|| Amtsgericht Stuttgart || HRB 746635
Geschäftsführer: Dipl.-Wi.-Ing. Bernhard Volz
USt-IdNr.: DE291682477
